Created matmul lab 1 for universities (#35456)
tenstorrent/tt-metal: docs/source/tt-metalium/tt_metal/labs/matmul/lab1/lab1.rst

for both transfers to complete. This is important because the kernel should not signal that the tiles are ready for consumption
until data is actually available.

The reader kernel repeats this process for all tiles. Given that circular buffers were created with two tiles each, the reader kernel can
read a new tile while the compute kernel is processing the previous one.

Compute Kernel Code
-------------------

The compute kernel in ``tt_metal/programming_examples/lab_eltwise_binary/kernels/compute/tiles_add.cpp`` is responsible for performing the elementwise addition of two tiles.

The function can be summarized by the following pseudo-code:

.. code-block:: cpp

   read_compile_time_arguments()
   initialize_tensix_engine_for_elementwise_addition()
   for (i in 0 .. n_tiles) {
       add_tiles_in_input_circular_buffers()
       write_result_to_output_circular_buffer()
   }

The kernel reads the number of tiles to process as a compile-time argument, enabling compiler optimizations such as loop unrolling.

An important architectural detail is that the compute kernel actually runs on three different RISC-V processors within the Tensix core:

> an unpacker, a compute processor, and a packer.

Reviewer: Here it gets a bit tricky, because we previously called them RISC-V 1-3, usually call them TRISC0-2, and colloquially UNPACK, MATH, PACK. Maybe a simple explanation will do here:
```suggestion
an unpacker (RISC-V 1 from Tensix Core image), a compute processor (RISC-V 2), and a packer (RISC-V 3).
```

The compiler automatically generates appropriate code for each of these three processors from the same source code, relieving the programmer
from having to write different code for each processor.
The unpacker handles reading data from circular buffers, the compute processor performs the actual arithmetic operations using the
Floating-Point Unit (FPU) of the Tensix engine, and the packer writes results back to circular buffers.
It is worth repeating that these RISC-V processors don't perform actual computations or packing/unpacking operations.
They simply issue commands to the Tensix engine to perform the actual computations and packing/unpacking operations.

The compute kernel uses circular buffers ``c_0`` and ``c_1`` for the two input tensors and ``c_16`` for the output tensor.
There are 32 circular buffers in total (0-31), and the exact indices used are up to programmer's discretion, provided they are used consistently
(i.e. reader and writer kernels must use the corresponding indices as the compute kernel).

The kernel initializes the FPU for elementwise binary operations, first calling ``binary_op_init_common``

> to set up the general binary operation infrastructure, followed by ``add_tiles_init`` to configure the FPU specifically for addition.

Reviewer: I don't feel strongly in any way, I just want to make sure that people involved know what to say when if a student asks them "What is meant by hardware infrastructure?".

This initialization only needs to be done once before the main loop, since all tiles use the same operation.

The main processing loop iterates over all tiles.
For each tile, the kernel first waits for one tile to become available in each input circular buffer using blocking calls to ``cb_wait_front``.

> These blocking calls ensure that the reader kernel has finished transferring the data before the compute kernel attempts to use it.

Reviewer: Not wrong, but a bit weird to speak from the position of the reader kernel. I would phrase it like:

```suggestion
These blocking calls ensure that the compute kernel doesn't attempt to use the data before the reader kernel has finished transferring it.
```

The kernel then acquires access to the destination register array using ``tile_regs_acquire`` and calls ``add_tiles`` to perform the elementwise addition.
The destination register array is a special storage area in the FPU that can hold multiple tiles and serves as the temporary
output location for FPU computations. The acquire operation also initializes all tiles in the destination register array to zero,
which is not important for our example, but is useful for operations like matrix multiplication where results accumulate.

After the computation completes, the kernel marks the input tiles as consumed using ``cb_pop_front`` to free space in the circular buffers,
then releases the destination register using ``tile_regs_commit`` to signal that the compute core has finished writing, which allows the packer to read the result.

The packer core for its part waits for the destination register to be ready using ``tile_regs_wait``,

> ensures there is space in the output circular buffer, and then copies the result from the destination register to the output circular buffer using ``pack_tile``.

Reviewer: ```suggestion
ensures there is space in the output circular buffer using ``cb_reserve_back``, and then copies the result from the destination register to the output circular buffer using ``pack_tile``.
```

Finally, it marks the output tile as ready using ``cb_push_back`` and releases the destination register using ``tile_regs_release``.

While it may seem like some of these operations are redundant (e.g. waiting on the destination register when it has seemingly just been released),
it is important to remember that compute kernel code is executed on three different RISC-V processors.
This synchronization mechanism using acquire, commit, wait, and release ensures that the three RISC processors coordinate properly,

> with the compute processor writing results and the packer processor reading them without conflicts.

Reviewer: nit:
```suggestion
with the compute processor writing results to the intermediate destination register array, and the packer processor reading them from it without conflicts.
```

  (``tt_metal/api/tt-metalium/tensor_accessor_args.hpp`` for host-side and ``tt_metal/hw/inc/api/tensor/tensor_accessor_args.h`` for device-side).

* Integers ``src0_addr``, ``src1_addr``, and ``dst_addr`` in the ``eltwise_add_tensix()`` function are host-side integers, but they contain addresses
  of the input and output tensor data in device DRAM.

* Kernel code and their arguments are JIT-compiled on the host, but then transferred to the device for execution.

Understanding the location of data and code on the host and device is useful when debugging or analyzing performance.

Kernel Compilation and Execution
********************************

As mentioned earlier, kernels are JIT compiled and executed on the device. This presents both advantages and disadvantages during development.
On the one hand, if one updates only kernel code, there is no need to rebuild before running the program to test that the changes had the desired effect.
On the other hand, it also means that errors in the kernel code will not be caught at host-code compile time, but only at time of host code execution,
when JIT compilation is triggered.

> Exercise 3: Observing JIT Compile Errors

Reviewer: Awesome!

.. figure:: images/matrix_3x4_row_major.jpg
   :width: 750
   :alt: A 3x4 Matrix A in Row-Major Layout
   :align: center

When accessing element ``A[i][j]`` in a matrix with ``N`` columns, the memory address is calculated as::

   address = base_address + (i * N + j) * sizeof(element)

Cache-Friendly Access Patterns
==============================

Modern CPUs fetch data in cache lines of consecutive words, so accessing nearby addresses is much faster than scattered accesses.
Accessing matrix elements in row-major order (left-to-right, top-to-bottom) is cache-friendly because consecutive elements in a row are already loaded in the cache.
Conversely, accessing matrix elements column-by-column is cache-unfriendly because each access may involve a different cache line.

Implications for Matrix Multiplication
--------------------------------------

In the standard matrix multiplication algorithm ``C = A * B``, where ``C[i][j] = ∑ₖ A[i][k] * B[k][j]``, and assuming ``i``, ``j``, ``k`` loop order:

* Accessing matrix ``A`` is cache-friendly because consecutive elements in memory are accessed one after another for two consecutive values of inner loop iterator ``k``.

> * Accessing matrix ``B`` is not cache-friendly because memory accesses skip a whole matrix row for two consecutive values of ``k``.

Reviewer: ```suggestion
* Accessing matrix ``B`` is cache-hostile because memory accesses skip a whole matrix row for two consecutive values of ``k``.
```

Assuming that the reader now focuses on the part of the phrasing after "cache-" to see if the described method is friendly or not, they might miss out completely that the negation came before the phrase itself, having to go back and read the sentence again. 

I know that I am maybe exaggerating here, but in this review I am assuming that the reader doesn't necessarily know English on a professional level, and might even use other tools to guide them through the text.

This asymmetry significantly impacts performance and motivates various optimization techniques such as loop reordering or tiling.

Linear Transformation and Computational Flexibility
***************************************************

Matrix multiplication represents a linear transformation where the computation can be viewed as a sequence of dot products.
This linearity allows significant computational flexibility, such as:

1. Loop Reordering: Changing the order of loops does not affect the result, but can impact performance on architectures with cache hierarchies.
2. Loop Tiling: A subset of the resulting matrix can be calculated at a time, improving cache locality and performance.

> 3. Parallelization: Different matrix regions can be computed concurrently.

Reviewer: ```suggestion
3. Parallelization: Different matrix regions can be computed simultaneously.
```

with explicit and carefully orchestrated data movement.
Typical workloads move input data from the host to device DRAM once, perform many computation steps directly on Tensix cores,
and transfer only final outputs back to the host.
When used this way, host-device communication over PCIe represents a relatively small fraction of the total runtime,
allowing the accelerator's on-device compute and data-movement engines to dominate performance.

In this lab, we will focus on programming a single Tensix core in TT-Metalium. Descriptions of the architectural features
will be limited to this context.
We will extend the architectural and programming model description to multiple Tensix cores in subsequent labs.

Tile-Based Architecture
=======================

The Tenstorrent architecture is a tile-based architecture.
This means that the data is processed in tiles, which are commonly ``32x32`` blocks of data.
Similar to vector architectures, which can perform an operation on one or more vectors of data using a single instruction,
a Tensix core can perform matrix operations on one or more tiles in one instruction.

> For example, a Tensix core can perform a matrix multiplication on two tiles each having ``32x32`` elements and produce a ``32x32`` result tile in one instruction.

Author: Many thanks for clarifying that. Reworded.

.. code-block:: cpp

   #include "api/debug/dprint.h"

   void kernel_main() {
       uint32_t iter = 5;
       bool done = false;

       // DPRINT uses a streaming syntax; ENDL() flushes the print buffer.
       DPRINT << "iter = " << iter << ENDL();

       // Booleans should be cast to an integer type before printing.
       DPRINT << "done flag = " << static_cast<uint32_t>(done) << ENDL();
   }

Printing a tile in a writer (dataflow) kernel
---------------------------------------------

Data is passed between kernels using circular buffers (CBs), which often contain tiles of data.
DPRINT can be combined with the ``TileSlice`` helper to print part or all of a tile from a CB.

The key rules are:

- You can only safely sample a tile from a CB **between** the appropriate CB API calls:

  - When reading from CBs (e.g. in writer kernels): between ``cb_wait_front()`` and ``cb_pop_front()``.
  - When writing to CBs (e.g. in reader kernels): between ``cb_reserve_back()`` and ``cb_push_back()``.

- The math (compute) RISC cannot directly see CBs, so CB tile printing is most commonly done from

>   the reader or writer kernels.

Reviewer 2: directly is a bit misleading. At the very least, tile printing works for me on the UNPACK thread in between cb_wait_front and cb_pop_front. I haven't tried other combinations, although many should work.

Reviewer 2: You do talk about it earlier.
Look for RISC-V 1 through RISC-V 3. MATH is RISC-V 2. 

I would just have one key rule and drop this line. If they try to use this without UNPACK/PACK/MATH, they'll just see a message instead of the tile for MATH.
If you really want to keep it, maybe say something like 
RISC-V 2 cannot directly see CBs, so CB tile data is only printed from other RISC-Vs.

Author: What I meant is that I don't discuss UNPACK/PACK/MATH specific ***macros*** (i.e. DPRINT_MATH, DPRINT_PACK,...)
I guess your point is that if they use regular DPRINT it will still execute on the other two cores, which could still be useful for them?
I guess I haven't thought of that. If that's the case I agree; I'll remove this statement.

  The data movement processors can handle complex access patterns while the compute units remain fully utilized.
* **Tiled operations**: The hardware is optimized for tiled operations, making tile-based algorithms essential for achieving peak performance.
  All matrices are processed in tile units, matching the natural granularity of the underlying hardware accelerators.
* **Pipelined data movement**: The circular buffer architecture with double buffering enables overlapped execution - while the compute kernel
  processes current tiles, the data movement kernels can simultaneously fetch the next set of tiles.
  This pipelining ensures efficient utilization of compute resources by minimizing idle time.

Troubleshooting and Additional Resources
****************************************

In rare cases, a Tensix device may enter an undefined operational state if a program performs actions outside the supported behavior.
In such a case, the ``tt-smi -r`` command can be used to reset the device.
This operation restores the device to a clean state, allowing normal operation to resume.

> If you encounter unexplained behaviors, try resetting the device using this command.

Author: I'm not sure that students will have the option to reboot the entire system, since it will be shared. Given this, do you still think it's worth mentioning here (rather than say in some general document about working with the system)?

> If you encounter unexplained behaviors, try resetting the device using this command.

Reviewer 2: Chances are this will happen at some point to someone.
